### PR TITLE
New highlighter based on tokens and untyped trees

### PR DESCRIPTION
**Pull request by allanrenucci**

### CODE REVIEW

**Comment by allanrenucci** [comment]
CLA is signed by all contributors:
- https://www.lightbend.com/contribute/cla/scala/check/slothspot
- https://www.lightbend.com/contribute/cla/scala/check/danielyli

The dotty bot is confused once again. @smarter this is ready for review

File: compiler/src/dotty/tools/dotc/printing/SyntaxHighlighting.scala [hunk 1]
```diff
@@ -22,343 +29,108 @@ object SyntaxHighlighting {
...
+          trees.foreach(traverse)
+
+        def traverse(tree: Tree)(implicit ctx: Context): Unit = {
+          tree match {
+            case tree: NameTree if ignored(tree) =>
+              ()
+            case tree: MemberDef /* ValOrDefDef | ModuleDef | TypeDef */ =>
+              for (annotation <- tree.rawMods.annotations)
+                highlightPosition(annotation.pos, AnnotationColor)
+              val color = if (tree.isInstanceOf[ValOrDefDef]) ValDefColor else TypeColor
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/printing/SyntaxHighlighting.scala]: Use pattern match instead of `isInstanceOf`

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/printing/SyntaxHighlighting.scala]: Maybe could just be another case in the outer pattern match and `for (annotation <- ...` could go in a helper method.

File: compiler/src/dotty/tools/repl/ReplDriver.scala [hunk 2]
```diff
@@ -284,7 +284,9 @@ class ReplDriver(settings: Array[String],
           x.symbol
       }
       .foreach { sym =>
-        out.println(SyntaxHighlighting("// defined " + sym.showUser))
+        // FIXME syntax highlighting on comment is currently not working
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/repl/ReplDriver.scala]: Will need to open an issue for this

**Pull request merged**